src/ol/layer/Group.js

```diff
@@ -205,19 +205,27 @@ class LayerGroup extends BaseLayer {
   }
 
   /**
-   * @param {Array<import("./Layer.js").State>} [opt_states] Optional list of layer states (to be modified in place).
+   * Get the layer states list and use this groups z-index as the default
+   * for all layers in this and nested groups, if it is unset at this point.
+   * If opt_states is not provided and this group's z-index is undefined
+   * 0 is used a the default z-index.
+   * @param {Array<import("./Layer.js").State>} [opt_states] Optional list
+   * of layer states (to be modified in place).
    * @return {Array<import("./Layer.js").State>} List of layer states.
    */
   getLayerStatesArray(opt_states) {
     const states = opt_states !== undefined ? opt_states : [];
-
     const pos = states.length;
 
     this.getLayers().forEach(function (layer) {
       layer.getLayerStatesArray(states);
     });
 
     const ownLayerState = this.getLayerState();
+    let defaultZIndex = ownLayerState.zIndex;
+    if (!opt_states && ownLayerState.zIndex === undefined) {
+      defaultZIndex = 0;
```

Comment:
It isn’t clear to me why this is necessary to solve the problem.

Reply:
> At the end of the `getLayerStatesArray` call chain all layer states should have a defined zIndex

Kind of subtle that `getLayerState` will return `zIndex: undefined` and `getLayerStates` will return `zIndex: 0`.  I don't have my head enough into this to know if this is what we want or not.  Just that it tripped me up when reviewing the test changes.